agent/consul/cluster_endpoint.go

```diff
@@ -0,0 +1,298 @@
+	"github.com/hashicorp/consul/lib"
+	"github.com/hashicorp/consul/lib/template"
+	"github.com/hashicorp/consul/tlsutil"
+	bexpr "github.com/hashicorp/go-bexpr"
+)
+
+type AutoConfigOptions struct {
+	NodeName    string
+	SegmentName string
+}
+
+type AutoConfigAuthorizer interface {
+	// Authorizes the request and returns a struct containing the various
+	// options for how to generate the configuration.
+	Authorize(*agentpb.AutoConfigRequest) (AutoConfigOptions, error)
+}
+
+type disabledAuthorizer struct{}
+
+func (_ *disabledAuthorizer) Authorize(_ *agentpb.AutoConfigRequest) (AutoConfigOptions, error) {
+	return AutoConfigOptions{}, fmt.Errorf("Auto Config is disabled")
+}
+
+type jwtAuthorizer struct {
+	validator       *ssoauth.Validator
+	allowReuse      bool
+	claimAssertions []string
+}
+
+func (a *jwtAuthorizer) Authorize(req *agentpb.AutoConfigRequest) (AutoConfigOptions, error) {
+	// perform basic JWT Authorization
+	identity, err := a.validator.ValidateLogin(context.Background(), req.JWT)
+	if err != nil {
+		// TODO (autoconf) maybe we should add a more generic permission denied error not tied to the ACL package?
+		return AutoConfigOptions{}, acl.PermissionDenied("Failed JWT authorization: %v", err)
+	}
+
+	varMap := map[string]string{
+		"node":    req.Node,
+		"segment": req.Segment,
+	}
+
+	// TODO (autoconf) check for JWT reuse if configured to do so.
+	for _, raw := range a.claimAssertions {
+		// validate and fill any HIL
+		filled, err := template.InterpolateHIL(raw, varMap, true)
+		if err != nil {
+			return AutoConfigOptions{}, fmt.Errorf("Failed to render claim assertion template %q: %w", raw, err)
+		}
+
+		evaluator, err := bexpr.CreateEvaluatorForType(filled, nil, identity.SelectableFields)
+		if err != nil {
+			return AutoConfigOptions{}, fmt.Errorf("Failed to create evaluator for claim assertion %q: %w", filled, err)
+		}
+
+		ok, err := evaluator.Evaluate(identity.SelectableFields)
+		if err != nil {
+			return AutoConfigOptions{}, fmt.Errorf("Failed to execute claim assertion %q: %w", filled, err)
+		}
+
+		if !ok {
+			return AutoConfigOptions{}, acl.PermissionDenied("Failed JWT claim assertion")
+		}
+	}
+
+	return AutoConfigOptions{
+		NodeName:    req.Node,
+		SegmentName: req.Segment,
+	}, nil
+}
+
+// Cluster endpoint is used for cluster configuration operations
+type Cluster struct {
+	srv *Server
+
+	authorizer AutoConfigAuthorizer
+}
+
+// updateTLSCertificatesInConfig will ensure that the TLS settings regarding how an agent is
+// made aware of its certificates are populated. This will only work if connect is enabled and
+// in some cases only if auto_encrypt is enabled on the servers. This endpoint has the option
+// to configure auto_encrypt or potentially in the future to generate the certificates inline.
+func (c *Cluster) updateTLSCertificatesInConfig(opts AutoConfigOptions, conf *config.Config) error {
+	if c.srv.config.ConnectEnabled && c.srv.config.AutoEncryptAllowTLS {
```

Comment:
Enabling `AutoEncryptAllowTLS`, automatically enables `ConnectEnabled`: https://github.com/hashicorp/consul/blob/a9851e181259f3e09240d1afee7a8406755851fd/agent/config/builder.go#L695-L699
Does AutoConfig explicitly rely on Connect? If it only needs that for AutoEncrypt, the check here is not needed imho.

Reply:
If the check is not needed, the content of this function would look like this:

```
conf.AutoEncrypt = &config.AutoEncrypt{TLS: c.srv.config.AutoEncryptAllowTLS}
return nil
```

Reply:
That depends on what you want AutoConfig to do for you.

You could have `verify_incoming_rpc` set to false on the servers while still having TLS certs. In that case AutoConfig can generate the agent token and push down some configuration without enabling auto encrypt to get TLS certificates for the Agent. There could be cases where this is still desired.

I could make this just check if AutoEncrypt.AllowTLS is enabled though.